apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnectionStringBuilder.cs

```diff
@@ -146,13 +146,35 @@ private string ValidatedDataSource(string dataSource)
         /// </summary>
         public ServiceEndpoint EndPoint => new ServiceEndpoint(Host, Port);
 
+        private InstanceHostManager _instanceHostManager = InstanceHostManager.Default;
+
+        /// <summary>
+        /// The <see cref="InstanceHostManager"/> to get host for instance.
+        /// </summary>
+        /// <remarks>
+        /// This property defaults to <see cref="InstanceHostManager.Default"/>
+        /// </remarks>
+        public InstanceHostManager InstanceHostManager
+        {
+            get => _instanceHostManager;
+            set => _instanceHostManager = GaxPreconditions.CheckNotNull(value, nameof(value));
+        }
+
         /// <summary>
-        /// The TCP Host name to connect to Spanner. If not supplied in the connection string, the default
-        /// host will be used.
+        /// The TCP Host name to connect to Spanner. If not supplied in the connection string, the random endpointUri 
+        /// available on instance or the default host will be used.
         /// </summary>
         public string Host
         {
-            get => GetValueOrDefault(nameof(Host), SpannerClient.DefaultEndpoint.Host);
+            get
+            {
+                if (ContainsKey(nameof(Host)))
+                {
+                    return (string)this[nameof(Host)];
+                }
+                string host = InstanceHostManager.GetHost(Project, SpannerInstance);
+                return host ?? SpannerClient.DefaultEndpoint.Host;
```

Comment:
This should set the host the first time it returns one to avoid returning a different one each time it is asked.
I think you can add a method similar to `GetOrAddValue(string, Func<string>)` and then use it here roughly like this:
`GetOrAddValue(nameof(Host), () => InstanceHostManager.GetHost(...)??SpannerClient.DefaultEnpoint.Host)`